lib/grandpa/grandpa.go

```diff
@@ -38,178 +50,192 @@ func NewService(blockState BlockState, voters []*Voter) (*Service, error) {
+		if err != nil {
+			return nil, err
+		}
+
+		if total >= uint64(2*len(s.state.voters)/3) {
+			var h *types.Header
+			h, err = s.blockState.GetHeader(pred)
+			if err != nil {
+				return nil, err
+			}
+
+			prevoted[pred] = uint64(h.Number.Int64())
+		} else {
+			prevoted, err = s.getPossiblePreVotedPredecessors(votes, pred, prevoted)
+			if err != nil {
+				return nil, err
+			}
+		}
 	}
 
-	return nil
+	return prevoted, nil
 }
 
-// checkForEquivocation checks if the vote is an equivocatory vote.
-// it returns true if so, false otherwise.
-// additionally, if the vote is equivocatory, it updates the service's votes and equivocations.
-func (s *Service) checkForEquivocation(voter *Voter, vote *Vote) bool {
-	v := voter.key.AsBytes()
-
-	if s.equivocations[v] != nil {
-		// if the voter has already equivocated, every vote in that round is an equivocatory vote
-		s.equivocations[v] = append(s.equivocations[v], vote)
-		return true
+// getTotalVotesForBlock returns the total number of observed votes for a block B, which is equal
+// to the direct votes for B and B's descendants plus the total number of equivocating voters
+func (s *Service) getTotalVotesForBlock(hash common.Hash) (uint64, error) {
+	// observed votes for block
+	dv, err := s.getVotesForBlock(hash)
+	if err != nil {
+		return 0, err
 	}
 
-	if s.votes[v] != nil {
-		// the voter has already voter, all their votes are now equivocatory
-		prev := s.votes[v]
-		s.equivocations[v] = []*Vote{prev, vote}
-		delete(s.votes, v)
-		return true
-	}
+	// equivocatory votes
+	ev := len(s.equivocations)
 
-	return false
+	return dv + uint64(ev), nil
 }
 
-// validateVote checks if the block that is being voted for exists, and that it is a descendant of a
-// previously finalized block.
-func (s *Service) validateVote(v *Vote) error {
-	// check if v.hash corresponds to a valid block
-	has, err := s.blockState.HasHeader(v.hash)
-	if err != nil {
-		return err
-	}
+// getVotesForBlock returns the number of observed votes for a block B.
+// The set of all observed votes by v in the sub-round stage of round r for block B is
+// equal to all of the observed direct votes cast for block B and all of the B's descendants
+func (s *Service) getVotesForBlock(hash common.Hash) (uint64, error) {
+	votes := s.getDirectVotes()
 
-	if !has {
-		return ErrBlockDoesNotExist
-	}
+	// B will be counted as in it's own subchain, so don't need to start with B's vote count
+	votesForBlock := uint64(0)
 
-	// check if the block is an eventual descendant of a previously finalized block
-	isDescendant, err := s.blockState.IsDescendantOf(s.head, v.hash)
-	if err != nil {
-		return err
-	}
+	for v, c := range votes {
 
-	if !isDescendant {
-		return ErrDescendantNotFound
-	}
+		// check if the current block is a descendant of B
+		isDescendant, err := s.blockState.IsDescendantOf(hash, v.hash)
+		if err != nil {
+			return 0, err
+		}
 
-	return nil
-}
+		if !isDescendant {
```

Comment:
this could be a else if, since its related to previous statement.

Reply:
same as above